### PR TITLE
fix: bytecode version unspecified & NPE and trySetAccessible method does not exist on Android

### LINKED ISSUE #1634: JavaAdapter unavailable in Android environment on upstream rhino
## Environment
- Android: 14
- rhino: 1.7.16-SNAPSHOT (based on b08de525f3daefa9800a509a5e5a5b27848da7ff commit)

## Code
```js
const timer = new java.util.TimerTask({ run: () => {} });
```

## What's expected
Android does not support the jvm's .class file format, so you need to use InterfaceAdapter to create a proxy and provide similar functionality instead of modifying the bytecode directly (works fine until rhino 1.7.15)

Screenshot_20240919_215216

## Actual behavior
throw this error

```
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  java.lang.ExceptionInInitializerError
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.JavaAdapter.createAdapterCode(JavaAdapter.java:328)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.JavaAdapter.getAdapterClass(JavaAdapter.java:312)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.JavaAdapter.js_createAdapter(JavaAdapter.java:169)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.JavaAdapter.execIdCall(JavaAdapter.java:87)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.IdFunctionObject.call(IdFunctionObject.java:85)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.BaseFunction.construct(BaseFunction.java:411)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.NativeJavaClass.construct(NativeJavaClass.java:173)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.Interpreter.interpretLoop(Interpreter.java:2052)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.Interpreter.interpret(Interpreter.java:1109)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:87)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:383)
2024-09-19 21:45:34.778 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:4029)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.javascript.InterpretedFunction.exec(InterpretedFunction.java:100)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at com.xfl.msgbot.script.api.modern.ScriptManager$Companion.initializeScript(ScriptManager.kt:261)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at com.xfl.msgbot.script.ScriptManagerCompat$Companion.initializeScript(ScriptManagerCompat.kt:151)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at com.xfl.msgbot.script.ScriptManagerCompat$Companion.initializeScript$default(ScriptManagerCompat.kt:137)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at com.xfl.msgbot.application.activity.ScriptListAdapter.onBindViewHolder$lambda$7$lambda$6(ScriptListAdapter.kt:151)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at com.xfl.msgbot.application.activity.ScriptListAdapter.$r8$lambda$-YWf-SZ-WLN1cmvdH5QnVuOnftM(Unknown Source:0)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at com.xfl.msgbot.application.activity.ScriptListAdapter$$ExternalSyntheticLambda5.run(D8$$SyntheticClass:0)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at java.lang.Thread.run(Thread.java:1012)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.io.InputStream.read(byte[], int, int)' on a null object reference
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	at org.mozilla.classfile.ClassFileWriter.<clinit>(ClassFileWriter.java:4376)
2024-09-19 21:45:34.779 12181-12447 System.err              com.xfl.msgbot                       W  	... 20 more
```

## The likely cause of the issue
Since the 1.7.15 release, there have been quite a few changes to the IRFactory and Parser. This is probably the cause.

### PR DESCRIPTION
**Pull request by naijun0403**
Fixes #1634

### CODE REVIEW

**Comment by gbrail** [comment]
Thanks -- I am going to open a PR soon that uses the "Errorprone" plugin to help detect these kinds of things proactively, but TBH nothing is going to be really permanently better until we have some way to do some sort of Android smoke test as part of the build, or somewhere in GitHub. Do you have any idea how other projects might handle this?

In the meantime, I'd love to understand better why the change to the class file writing is necessary. I'd be even happier if we could find some way to write a test for it -- any ideas?

**Comment by naijun0403** [comment]
Thank you for your feedback!

Regarding testing:
I'm aware that the nowinandroid repository uses Android emulators for testing. However, my knowledge of testing methodologies is limited, so I can't provide comprehensive suggestions for alternative testing methods.

Reason for modifying ClassFileWriter.java:
The modification was necessary due to Android's unique runtime environment. Android doesn't use the JVM, so .class files are not included in the app. Consequently, in the source code, inputStream was always null, leading to a NPE that prevented the entire engine from functioning.

Changes made:
1. I added a null check for inputStream.
2. If inputStream is null, I substitute default minor and major version values (retrieved from a previous commit) to ensure null safety in subsequent operations.

This approach aims to prevent NPE and maintain the engine's functionality in the Android environment.

File: rhino/src/main/java/org/mozilla/javascript/JavaMembers.java [hunk 1]
```diff
@@ -328,7 +328,8 @@ private void discoverAccessibleMethods(
                                 if (isPublic(mods) || isProtected(mods) || includePrivate) {
                                     MethodSignature sig = new MethodSignature(method);
                                     if (!map.containsKey(sig)) {
-                                        if (includePrivate) method.trySetAccessible();
+                                        if (includePrivate && !method.isAccessible())
+                                            method.setAccessible(true);
```
Comment (rPraml) [inline rhino/src/main/java/org/mozilla/javascript/JavaMembers.java:332]: I stumbled upon this place in #1575 (which probably will cause merge conflicts)
I'm wondering, if we should replace this with `VMBridge.instance.tryToMakeAccessible(method)`
which will effectively call this code: https://github.com/mozilla/rhino/blob/master/rhino/src/main/java/org/mozilla/javascript/jdk18/VMBridge_jdk18.java#L57

Comment (naijun0403) [reply rhino/src/main/java/org/mozilla/javascript/JavaMembers.java:332]: Maybe that's a good idea. I guess it will work on Android too.

**Pull request closed without merging**